jekyll/_posts/2021-03-12-new-Nim-channels.md

```diff
@@ -0,0 +1,208 @@
+
+## Background
+
+A channel is a model for sharing memory via message passing. A thread is able to send or receive messages over a channel. It's like sending a letter to your friend: the postman is the channel, and your friend is the receiver. You might already know `system/channels_builtin`, which is the old channels implementation. What's better in the new implementation? With the old one, first you need to copy your letter and send the copy to your friend. Then your friend may mark something on the copied letter and it won't affect the original. This works fine, but it is not efficient. If you use the new implementation, you only need to put your letter in the mailbox. No need to copy it!
+
+## The advantages
+
+- Designed for ARC/ORC, no legacy code
+- No need to `deepCopy`, just move data around
+- No data races
+
+## Explore the new channels
+
+**Note:** Be sure to compile your code with `--gc:orc –-threads:on -d:ssl`.
+
+### Let's crawl the web
+
+**todo_urls.json**
+
+```json
+{"url": ["https://google.com", "https://nim-lang.org"]}
+```
+
+**app.nim**
+
+The main thread prepares tasks by reading `todo_urls.json`, and then it sends JSON data to a channel. The crawl thread does the actual work - it receives URL data from the channel and downloads the contents using the `httpclient` module.
+
+```nim
+import std/channels
+import std/[httpclient, isolation, json]
+var ch = newChannel[JsonNode]() # we need to send JsonNode
+proc download(client: HttpClient, url: string) =
+  let response = client.get(url)
+  echo "content: ", response.body[0 .. 20] # prints the results
+proc crawl =
+  var client = newHttpClient() # the crawler
+  defer: client.close()
+  let data = ch.recv() # the JSON data
+  if data != nil:
+    for url in data["url"]:
+      download(client, url.getStr)
+proc prepareTasks(fileWithUrls: string): seq[Isolated[JsonNode]] =
+  result = @[]
+  for line in lines(fileWithUrls):
+    result.add isolate(parseJson(line)) # parse JSON file
+proc spawnCrawlers =
+  var tasks = prepareTasks("todo_urls.json")
+  for t in mitems tasks: # we need a mutable view of the items
+    ch.send move t
+var thr: Thread[void]
+createThread(thr, crawl) # create crawl thread
+spawnCrawlers()
+joinThread(thr)
+```
+
+First, we import `std/channels`.
+
+Then we can create a channel using `newChannel`, which returns a `Channel[T]`. It uses MPMC internally, which stands for "multiple producer, multiple consumer". The `elements` parameter is used to specify whether a channel is buffered or not. For an unbuffered channel, the sender and the receiver block until the other side is ready. Sending data to a buffered channel blocks only when the buffer is full. Receiving data from a buffered channel blocks when the buffer is empty.
+
+`newChannel` is a generic proc - you can specify the types of the data you want to send or receive.
```

Comment:
It would be nice to provide `newChannel` declaration here and expand the explanation a bit, since the mentioned parameter is not used in the text previously. Also, I've lifted the note on genericity up to the declaration description paragraph. Also, I don't think this is the right place to mention MPMC as it breaks the flow of folowing the code example above. It's probably better to mention it in the Background section.

````suggestion
Then we can create a channel using `newChannel` proc with the following signature:
```nim
proc newChannel*[T](elements = 30): Channel[T]
```
The new Nim channel is based on MPMC queue architecture internally, which stands for "multiple producer, multiple consumer". As you can see, `newChannel` is a generic proc - you can specify the type of the data you want to send or receive. The proc returns a `Channel[T]` and takes a single parameter `elements` with 30 as the default and which is used to specify whether a channel is buffered (`elements` = 1) or not (`elements` > 1). For an unbuffered channel, the sender and the receiver block until the other side is ready. Sending data to a buffered channel blocks only when the buffer is full. Receiving data from a buffered channel blocks when the buffer is empty.

Here's three examples of initializing a channel:
````

Reply:
thanks!